### DISCUSSION TITLE
Admin UI and `--hostname-admin` can't find `openid-connect/3p-cookies/step1.html`

### DISCUSSION THREAD

**Discussion by kou** [keycloak/keycloak#33441] [Archive]
### Before reporting an issue

- [X] I have read and understood the above terms for submitting issues, and I understand that my issue may be closed without action if I do not follow them.

### Area

admin/ui

### Describe the bug

My usage may be wrong but it seems that we can't use admin UI with `--hostname-admin`.

Run Keycloak with `--hostname-admin http://localhost:8080`:

```bash
podman \
  run \
    --rm \
    -p 127.0.0.1:8080:8080 \
    -p 127.0.0.1:9000:9000 \
    -e KEYCLOAK_ADMIN=admin \
    -e KEYCLOAK_ADMIN_PASSWORD=admin \
    quay.io/keycloak/keycloak:25.0.6 \
    start \
    --hostname http://localhost:9000 \
    --hostname-admin http://localhost:8080 \
    --http-enabled true
```

Open http://localhost:8080/ . It shows "somethingWentWrong" dialog.

Web browser's network debug console shows that http://localhost:9000/realms/master/protocol/openid-connect/3p-cookies/step1.html (see the port number: `9000`) returns 404.

Note that http://localhost:8080/realms/master/protocol/openid-connect/3p-cookies/step1.html (see the port number: `8080`) returns 200.

### Version

25.0.6

### Regression

- [ ] The issue is a regression

### Expected behavior

We can use admin UI with `--hostname-admin`.

### Actual behavior

http://localhost:8080/ shows "somethingWentWrong" dialog.

### How to Reproduce?

Run Keycloak with `--hostname-admin`:

```bash
podman \
  run \
    --rm \
    -p 127.0.0.1:8080:8080 \
    -p 127.0.0.1:9000:9000 \
    -e KEYCLOAK_ADMIN=admin \
    -e KEYCLOAK_ADMIN_PASSWORD=admin \
    quay.io/keycloak/keycloak:25.0.6 \
    start \
    --hostname http://localhost:9000 \
    --hostname-admin http://localhost:8080 \
    --http-enabled true
```

Open http://localhost:8080/ .

### Anything else?

We may need to use `--hostname-admin` URL not `--hostname` URL for `openid-connect/3p-cookies/step1.html`.

**Comment by shawkins** [comment]
Converted to a discussion for initial clarification.

> --hostname http://localhost:9000 \
> --hostname-admin http://localhost:8080 \

You set these when you are accessing via a proxy - see https://www.keycloak.org/server/hostname

These are telling keycloak to expect proxy access for the application / frontend via http://localhost:9000 and the admin via http://localhost:8080 so when returning links to the browser for the frontend or admin access it will use these.  These settings do not change the actual host / port / protocol that the keycloak server is listening on. 

Since there does not appear to be a proxy in use, you don't need these settings yet.

> Web browser's network debug console shows that http://localhost:9000/realms/master/protocol/openid-connect/3p-cookies/step1.html (see the port number: 9000) returns 404.

The default management port is 9000, so you are actually accessing that, not the application. 

You have not changed the http-port, so the keycloak application is listening to http traffic at port 8080.

> We may need to use --hostname-admin URL not --hostname URL for openid-connect/3p-cookies/step1.html.

Any url that does not have a relative root of /admin is part of the front end of keycloak and should be accessed via the public hostname, not the hostname-admin.

**Reply by kou** [reply]
Thanks for your reply.

Sorry. I simplified how to reproduce too much. This is also happen with reverse proxy.

nginx configuration:

```text
server {
	listen 80;
	listen [::]:80;
	server_name id.localhost;

	proxy_set_header Host $host;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Host $host;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header X-Forwarded-Port 80;

	location / {
		proxy_pass http://localhost:9000/;
	}
}

server {
	listen 80;
	listen [::]:80;
	server_name admin.id.localhost;

	proxy_set_header Host $host;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Host $host;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header X-Forwarded-Port 80;

	location / {
		proxy_pass http://localhost:8080/;
	}
}
```

Add the following to `/etc/hosts` to use `id.localhost` and `admin.id.localhost`:

```text
127.0.0.1	id.localhost admin.id.localhost
```

Keycloak:

```bash
podman \
  run \
  --rm \
  -p 127.0.0.1:8080:8080 \
  -p 127.0.0.1:9000:9000 \
  -e KEYCLOAK_ADMIN=admin \
  -e KEYCLOAK_ADMIN_PASSWORD=admin \
  quay.io/keycloak/keycloak:25.0.6 \
  start \
  --hostname http://id.localhost \
  --hostname-admin http://admin.id.localhost \
  --http-enabled true \
  --proxy-headers xforwarded
```

Open http://admin.id.localhost/ .

It requests http://id.localhost/realms/master/protocol/openid-connect/3p-cookies/step1.html (`id.localhost` not `admin.id.localhost`) and returns 404.

http://admin.id.localhost/realms/master/protocol/openid-connect/3p-cookies/step1.html (`admin.id.localhost` not `id.localhost`) returns 200.

**Reply by shawkins** [reply]
> and returns 404

You are still trying to access the management port 9000 (for metrics and other functionality) not the keycloak application with "proxy_pass http://localhost:9000/;"

This should instead be using port 8080.

**Reply by kou** [reply]
Ah, thanks. I understand. I need to use `localhost:8080` for both `hostname` and `hostname-admin`:

```text
server {
	listen 80;
	listen [::]:80;
	server_name id.localhost;

	proxy_set_header Host $host;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Host $host;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header X-Forwarded-Port 80;

	location / {
		proxy_pass http://localhost:8080/;
	}
}

server {
	listen 80;
	listen [::]:80;
	server_name admin.id.localhost;

	proxy_set_header Host $host;
	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
	proxy_set_header X-Forwarded-Host $host;
	proxy_set_header X-Forwarded-Proto $scheme;
	proxy_set_header X-Forwarded-Port 80;

	location / {
		proxy_pass http://localhost:8080/;
	}
}
```

I misunderstood that I need to use separated host names for `localhost:8080` and `localhost:9000`.

Can I open a PR that updates documents to clarify this?
If so, should I update https://www.keycloak.org/server/hostname#_exposing_the_administration_console_on_a_separate_hostname or https://www.keycloak.org/server/reverseproxy (or both of them)?

**Comment by shawkins** [comment]
> Can I open a PR that updates documents to clarify this?

PRs are always welcome. Anything that lacks clarity should certainly be addressed.

**Reply by kou** [reply]
OK. How about #33550?
